### PR TITLE
storage: Use btrfs subvolume list command instead of walking dir to get subvolumes

### PR DESCRIPTION
**Pull request by takonomura**
This pull request changes `getSubvolumes` to use command `btrfs subvolume list` instead of walking the entire tree.

It took over three seconds to walk the entire tree in my test environment with a large tree. This can be reduced to a few milliseconds using the command.

### CODE REVIEW

File: lxd/storage/drivers/driver_btrfs_utils.go [hunk 1]
```diff
@@ -91,38 +91,39 @@ func (d *btrfs) isSubvolume(path string) bool {
...
 		path = path + "/"
 	}
 
-	// Walk through the entire tree looking for subvolumes.
-	err := filepath.WalkDir(path, func(fpath string, entry fs.DirEntry, err error) error {
-		if err != nil {
-			return err
-		}
+	var stdout bytes.Buffer
+	err := shared.RunCommandWithFds(context.TODO(), nil, &stdout, "btrfs", "subvolume", "list", poolMountPath)
```
Comment (monstermunchkin) [inline lxd/storage/drivers/driver_btrfs_utils.go:107]: Please use `d.state.ShutdownCtx` instead of `context.TODO()`.

**Review by monstermunchkin — DISMISSED** [review]

File: lxd/storage/drivers/driver_btrfs_utils.go [hunk 2]
```diff
@@ -91,38 +91,39 @@ func (d *btrfs) isSubvolume(path string) bool {
 }
 
 func (d *btrfs) getSubvolumes(path string) ([]string, error) {
-	result := []string{}
+	poolMountPath := GetPoolMountPath(d.name)
```
Comment (tomponline) [inline lxd/storage/drivers/driver_btrfs_utils.go:94]: This check should be in a separate commit as isn't related to the actual change described in the commit message.

This makes it easier to backport to LTS series.

File: lxd/storage/drivers/driver_btrfs_utils.go [hunk 3]
```diff
@@ -91,38 +91,39 @@ func (d *btrfs) isSubvolume(path string) bool {
...
 		path = path + "/"
 	}
 
-	// Walk through the entire tree looking for subvolumes.
-	err := filepath.WalkDir(path, func(fpath string, entry fs.DirEntry, err error) error {
-		if err != nil {
-			return err
-		}
+	var stdout bytes.Buffer
+	err := shared.RunCommandWithFds(d.state.ShutdownCtx, nil, &stdout, "btrfs", "subvolume", "list", poolMountPath)
```
Comment (tomponline) [inline lxd/storage/drivers/driver_btrfs_utils.go:107]: Do you have an insight into how this command works? I'm wondering if it does anything different than the current implementation in this function does? Does it have access to btrfs internals that makes this operation faster?

Comment (tomponline) [reply lxd/storage/drivers/driver_btrfs_utils.go:107]: Ah I just saw:

> It took over three seconds to walk the entire tree in my test environment with a large tree. This can be reduced to a few milliseconds using the command.

Thanks!

Comment (tomponline) [reply lxd/storage/drivers/driver_btrfs_utils.go:107]: I suppose it does have a more efficient way to do this then.

**Review by tomponline — APPROVED** [review]
LGTM, as we should be able to backport this in one lump there's no specific need in this case to split the commits.
